Decouple stats keeper from openvpn

vkuznecovas opened the pull request:
Updates #530

tadovas commented on client/stats/sender.go:
```diff
@@ -30,6 +29,16 @@ import (
+
+const (
+	// Connected represents the connected state
+	Connected State = iota
```
What is wrong with constants defined in connection manager?
Stats sender is aware and handles same connection states

tadovas commented on cmd/di.go:
```diff
@@ -213,13 +213,18 @@ func (di *Dependencies) bootstrapServiceOpenvpn(nodeOptions node.Options) {
 	}
 	di.ServiceRegistry.Register(service_openvpn.ServiceType, createService)
 
+	statsSenderFactory := func(statsKeeper stats.SessionStatsKeeper, mysteriumClient server.Client, sessionID session.ID, providerID identity.Identity, signer identity.Signer, consumerCountry string, interval time.Duration) service_openvpn.StatsSender {
```
Why stats sender has to be factory?

tadovas commented on services/openvpn/connection.go:
```diff
@@ -51,3 +52,16 @@ func GetStateCallback(stateChannel connection.StateChannel) func(openvpnState op
 }
+
+// GetStatsSenderCallback returns a stats sender callback that knows how to work with openvpn state
+func GetStatsSenderCallback(statsHandler func(state stats.State)) func(openvpnState openvpn.State) {
```
This seems to be overcomplicated and not necessary

tadovas commented on services/openvpn/connection_factory.go:
```diff
@@ -75,19 +94,20 @@ func (op *ProcessBasedConnectionFactory) newBytecountMiddleware() management.Mid
 func (op *ProcessBasedConnectionFactory) newStateMiddleware(session session.ID, signer identity.Signer, connectionOptions connection.ConnectOptions, stateChannel connection.StateChannel) management.Middleware {
-	// TODO Move stats.Sender from 'openvpn' package
-	statsSender := stats.NewRemoteStatsSender(
+	statsSender := op.statsSenderFactory(
```
Why that, statistics sender is leaked into openvpn service, and cannot be simply a part of conn manager? Each connection (a.k.a openvpn or other) should provide clear callbacks - stats, events etc. Conn manager handles them appropriatly

tadovas commented on services/openvpn/connection_factory_test.go:
```diff
@@ -76,3 +79,21 @@ func TestConnectionFactory_CreatesConnection(t *testing.T) {
 	assert.NotNil(t, conn)
 }
+
+type FakeStatsSender struct {
```
Mock or stub is more appropriate definition in test world

vkuznecovas commented:
I've removed the stats sender from openvpn completely. Moved it up to connection manager. The stats sender now only has a start and a stop method, it knows absolutely nothing about states.

tadovas commented on core/connection/stub_stat_sender.go:
```diff
@@ -0,0 +1,34 @@
+package connection
+
+// StubStatSender represents a stat sender stub
+type StubStatSender struct {
```
I guess it's used in tests? Does it really deservers it's own file?

tadovas commented on core/connection/manager.go:
```diff
@@ -166,6 +185,8 @@ func (manager *connectionManager) startConnection(consumerID identity.Identity,
 		Proposal:      proposal,
 	}
 
+	manager.statsSender = manager.statsSenderFactory(consumerID, sessionID, proposal, time.Minute)
```
This stats sender lifecycle is still strage to me. It's constructed on every connection, then on state callback its start/stop methods are called. So its temporal depedency - created on one method, methods called on different. Can that stats sender be made like component injected into connection manager and simply handle stats, and conn start stop .methods internally (by resetting itself when needed or doing separate unrelated goroutines whatever)

vkuznecovas replied:
I agree that it's strange. I think this component shouldn't even be a component, a func would do here. There's no state it needs to keep internally - just send a request with the given data. Maybe it needs to be merged with the keeper, just adding a .Send() to it and renaming it to a stat handler? I don't feel that the schedule part of the component makes it a worthy component at all. I'm also not entirely sure that this is the appropriate PR to rewrite this whole component in general - the idea here was to move it up a level, instead of having it under the openvpn connection.

vkuznecovas commented:
#567 includes this now, so i'll close it.
